### PR TITLE
ECP integration

### PR DESCRIPTION
**Pull request by lcforges**

### CODE REVIEW

**Comment by clundin25** [comment]
nit: Try to squash all of your commits into one before the initial review. That way it is easy to follow the evolution of the PR.

File: pkcs11/pkcs11.go [hunk 1]
```diff
@@ -1721,13 +1721,14 @@ func (r *rsaPrivateKey) getHash() *crypto.Hash {
 		return r.hash
 	}
 	// Initialized to SHA256 if no hash specified
```
Comment (clundin25) [inline pkcs11/pkcs11.go]: You should delete this comment, it is already out of sync with the code ;)

File: pkcs11/pkcs11.go [hunk 2]
```diff
@@ -1711,7 +1707,11 @@ func (s *Slot) generateECDSA(o keyOptions) (crypto.PrivateKey, error) {
 	return priv, nil
 }
 
-func (r *rsaPrivateKey) WithHash(hash crypto.Hash) (*rsaPrivateKey, error) {
+func WithHash(privKey crypto.PrivateKey, hash crypto.Hash) (crypto.PrivateKey, error) {
+	r, ok := privKey.(*rsaPrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("expected type rsaPrivateKey, but got type %T",privKey)
```
Comment (clundin25) [inline pkcs11/pkcs11.go]: I'm wondering if you should have a separate PR that runs `go fmt` on all the files touched. There seems to be a missing space before `privKey`.

File: pkcs11/pkcs11.go [hunk 3]
```diff
@@ -1855,3 +1856,20 @@ func makeCParamRSAOAEP(hash *crypto.Hash) (C.CK_RSA_PKCS_OAEP_PARAMS_PTR, error)
...
+	}
+	return rsaPriv.decryptRSA(encryptedData)
+}
+
+func Encrypt(key crypto.PrivateKey, data []byte) ([]byte, error) {
+	rsaPriv, ok := key.(*rsaPrivateKey) 
+	if !ok {
+		return nil, fmt.Errorf("Encrypt Error: expected type rsaPrivateKey, got %T", key)
+	}
+	return rsaPriv.encryptRSA(data)
```
Comment (clundin25) [inline pkcs11/pkcs11.go:1874]: This code would be generic if there was an `Encryptor` interface (Or something similar).

As a starting point:

```golang
type interface PKCS11Encryptor {
   func Encrypt(plaintext []byte) ([]byte, err)
}
```

File: pkcs11/pkcs11_test.go [hunk 4]
```diff
@@ -645,11 +648,14 @@ func TestDecryptRSA(t *testing.T) {
 			if err != nil {
 				t.Fatalf("generate(%#v) failed: %v", o, err)
 			}
+			priv, err = WithHash(priv, crypto.SHA1)
```
Comment (clundin25) [inline pkcs11/pkcs11_test.go:652]: Should keep the deleted comment.

File: pkcs11/pkcs11.go (hunk 1 above)
Comment (lcforges) [reply pkcs11/pkcs11.go]: Ahh, I meant to change line 1724 back to sha256. I've just updated the code to make the comment relevant again.

File: pkcs11/pkcs11.go (hunk 2 above)
Comment (lcforges) [reply pkcs11/pkcs11.go]: I committed gofmt changes on this PR, if that is okay.

Comment (clundin25) [reply pkcs11/pkcs11.go]: In the future it should be an isolated commit. Otherwise it makes it hard to see what changes are formatting vs real change

Comment (lcforges) [reply pkcs11/pkcs11.go]: Ahh okay, that makes sense

File: pkcs11/pkcs11.go [hunk 5]
```diff
@@ -1590,11 +1590,7 @@ func (s *Slot) generateRSA(o keyOptions) (crypto.PrivateKey, error) {
 	if err != nil {
 		return nil, fmt.Errorf("parsing private key: %w", err)
 	}
-	rsaPriv, ok := priv.(*rsaPrivateKey)
-	if !ok {
-		return nil, fmt.Errorf("expected rsa private key, got: %T", pub)
-	}
-	rsaPriv.WithPublicKeyHandle(pubObj)
+	WithPublicKeyHandle(priv, pubObj)
```
Comment (clundin25) [inline pkcs11/pkcs11.go]: Should `priv` be assigned to the result of this function?

**Review by clundin25 — DISMISSED** [review]
Please resolve my last comment then go ahead and merge

File: pkcs11/pkcs11.go (hunk 5 above)
Comment (lcforges) [reply pkcs11/pkcs11.go]: Yes, I think you are correct. I've changed it

**Pull request merged**